### PR TITLE
Split eigenvalue solve from eigenvalue selection and add choice of solvers

### PR DESCRIPTION
**Pull request by markdewing**
Move the eigenvalue selection code to a separate function, so it can be re-used with different solvers.

Add an "ev_solver" parameter to allow switching between the two implemented solution methods.
The choice of eigenvalue solvers is only implemented for the "one shift only" optimizer in the batched driver.

Prior to this PR, the code using the generalized eigensolver existed in the code, but was not connected.   More importantly, these changes are paving the way for a larger selection of eigenvalue solvers.

One change in logic is the eigenvalue problem is solved on one rank and the resulting change in parameters is broadcast to the other ranks. Previously, the eigenvalue problem was solved on each rank separately.  This is looking forward to different solvers which may not produce exactly the same answer when run multiple times.  The amount of data broadcast is small, so the effect on performance should be minimal.
The broadcast is here:
https://github.com/QMCPACK/qmcpack/blob/355302fe9da8dbb73e555b09b4b9f20ade6d5688/src/QMCDrivers/WFOpt/QMCFixedSampleLinearOptimizeBatched.cpp#L1693

I can add documentation (and more tests) if the code is okay.

## What type(s) of changes does this code introduce?
_Delete the items that do not apply_

- Refactoring (no functional changes, no api changes)

### Does this introduce a breaking change?

- No

## What systems has this change been tested on?
desktop

## Checklist

_Update the following with a yes where the items apply. If you're unsure about any of them, don't hesitate to ask.  This is
simply a reminder of what we are going to look for before merging your code._

- Yes. This PR is up to date with current the current state of 'develop'
- Yes. Code added or changed in the PR has been clang-formatted
- Yes. This PR adds tests to cover any new code, or to catch a bug that is being fixed
- Not yet. Documentation has been added (if appropriate)

### CODE REVIEW

File: src/QMCDrivers/WFOpt/QMCFixedSampleLinearOptimizeBatched.cpp [hunk 1]
```diff
@@ -1649,28 +1655,24 @@ bool QMCFixedSampleLinearOptimizeBatched::one_shift_run()
...
-
-    // transpose the result (why?)
-    for (int i = 0; i < N; i++)
-      for (int j = i + 1; j < N; j++)
-        std::swap(prdMat(i, j), prdMat(j, i));
+    RealType lowestEV;
+    // compute the lowest eigenvalue and the corresponding eigenvector
+    if (ev_solver_ == "general") {
+      app_log() << "  Using generalized eigenvalue solver (ggev)" << std::endl;
+      lowestEV = getLowestEigenvector_Gen(hamMat, invMat, parameterDirections);
```
Comment (ye-luo) [inline src/QMCDrivers/WFOpt/QMCFixedSampleLinearOptimizeBatched.cpp:1667]: > The code using the generalized eigensolver has been in the code, but has not been connected. 

Is it not connected here?

File: src/QMCDrivers/WFOpt/QMCFixedSampleLinearOptimizeBatched.cpp [hunk 2]
```diff
@@ -576,6 +576,11 @@ void QMCFixedSampleLinearOptimizeBatched::process(xmlNodePtr q)
   m_param.add(OutputMatricesHDF, "output_matrices_hdf", {"no", "yes"});
   m_param.add(FreezeParameters, "freeze_parameters", {"no", "yes"});
 
+  m_param.add(ev_solver_, "ev_solver",
```
Comment (ye-luo) [inline src/QMCDrivers/WFOpt/QMCFixedSampleLinearOptimizeBatched.cpp:579]: Would like to call it `eigensolver` than `ev_solver`.

File: src/QMCDrivers/WFOpt/QMCFixedSampleLinearOptimizeBatched.cpp (hunk 1 above)
Comment (markdewing) [reply src/QMCDrivers/WFOpt/QMCFixedSampleLinearOptimizeBatched.cpp:1667]: In the past the generalized eigensolver was not connected, but it is connected now.

Comment (ye-luo) [reply src/QMCDrivers/WFOpt/QMCFixedSampleLinearOptimizeBatched.cpp:1667]: I see what you mean. PR description reflects the code in the branch. So "has not been connected" confused me.

Comment (markdewing) [reply src/QMCDrivers/WFOpt/QMCFixedSampleLinearOptimizeBatched.cpp:1667]: I updated the PR description to be more clear.

File: src/QMCDrivers/WFOpt/QMCFixedSampleLinearOptimizeBatched.cpp (hunk 2 above)
Comment (ye-luo) [reply src/QMCDrivers/WFOpt/QMCFixedSampleLinearOptimizeBatched.cpp]: Would you like to add the manual entry in this PR or later? If later, I would like to approve the current PR.

Comment (markdewing) [reply src/QMCDrivers/WFOpt/QMCFixedSampleLinearOptimizeBatched.cpp]: Either way is fine. If you want to keep this moving, I can add the manual entry later.

**Review by ye-luo — APPROVED** [review]
Still need non-ANL approval

**Review by prckent — APPROVED** [review]
LGTM (airport review)

**Comment by prckent** [comment]
> Previously, the eigenvalue problem was solved on each rank separately.
I consider the new implementation a bug fix -- one day we would have been caught by the old implementation. DFT codes that have done similar things have run into trouble historically.

**Comment by prckent** [comment]
Test this please

**Comment by ye-luo** [comment]
> > Previously, the eigenvalue problem was solved on each rank separately.
> > I consider the new implementation a bug fix -- one day we would have been caught by the old implementation. DFT codes that have done similar things have run into trouble historically.

@markdewing could you post a link to the line of broadcast after solving on rank 0?

**Comment by markdewing** [comment]
I updated the PR description with a link to the broadcast, also given here:
https://github.com/QMCPACK/qmcpack/blob/355302fe9da8dbb73e555b09b4b9f20ade6d5688/src/QMCDrivers/WFOpt/QMCFixedSampleLinearOptimizeBatched.cpp#L1693

**Pull request merged**